File: apps/builder-mac/Sources/FlagshipBuilderCore/VM/VMLifecycle.swift
import Foundation

/// One hosted VM's lifecycle — pure and event-driven. The host (VZ adapter /
/// UI) feeds it events and executes the effects it returns; nothing in here
/// touches Virtualization.framework, the filesystem, or a real clock.
///
///     created ──startInstall──▶ installing (installer ISO attached)
///     installing ──installSucceeded──▶ installed (ISO DETACHED — from here
///                                       the guest boots from its own disk)
///     installed/stopped ──powerOn──▶ awaitingPhoneUnlock   (encrypted guest:
///                                       sealed in the initramfs until the
///                                       phone-home unlock answers)
///                              └────▶ running              (unencrypted guest)
///     awaitingPhoneUnlock ──guestUnlocked──▶ running
///     running/awaitingPhoneUnlock ──powerOff──▶ stopped
///     + failure states (install / run), each retryable.

public struct VMFailure: Codable, Sendable, Equatable {
    public enum Phase: String, Codable, Sendable {
        case install
        case run
    }
    public let phase: Phase
    public let reason: String

    public init(phase: Phase, reason: String) {
        self.phase = phase
        self.reason = reason
    }
}

public enum VMState: Codable, Sendable, Equatable {
    case created
    case installing
    case installed
    case awaitingPhoneUnlock
    case running
    case stopped
    case failed(VMFailure)

    /// User-facing status label (sidebar / detail).
    public var label: String {
        switch self {
        case .created: return "Created"
        case .installing: return "Installing…"
        case .installed: return "Installed"
        case .awaitingPhoneUnlock: return "Waiting for you to unlock"
        case .running: return "Running"
        case .stopped: return "Stopped"
        case .failed(let f):
            switch f.phase {
            case .install: return "Install failed"
            case .run: return "Stopped unexpectedly"
            }
        }
    }
}

public enum VMEvent: Sendable, Equatable {
    /// Begin (or retry) the unattended install from the remastered ISO.
    case startInstall
    /// The installer ran to completion (the guest powered itself off).
    case installSucceeded
    case installFailed(String)
    /// Boot the installed guest from its main disk.
    case powerOn
    /// The sealed guest's phone-home unlock completed (owner approved / a
    /// lease answered) and the guest is up.
    case guestUnlocked
    /// The guest stopped (user action or a clean guest shutdown).
    case powerOff
    case runtimeFailed(String)
}

/// Side effects the caller must perform, in order. The state machine decides;
/// the VZ layer obeys.
public enum VMEffect: Sendable, Equatable {
    case attachInstallerISO
    case detachInstallerISO
    case startVirtualMachine
    case stopVirtualMachine
}

public enum VMLifecycleError: Error, Equatable {
    case invalidTransition(from: VMState, on: VMEvent)
}

public struct VMLifecycle: Sendable {
    public private(set) var state: VMState
    /// When the current state was entered (via the injected clock).
    public private(set) var stateChangedAt: Date

    /// From VMConfig.awaitsPhoneUnlockAtBoot: whether powering on an installed
    /// guest lands in the sealed awaiting-phone-unlock state first.
    public let sealedAtBoot: Bool
    private let clock: @Sendable () -> Date

    public init(state: VMState = .created,
                sealedAtBoot: Bool,
                clock: @escaping @Sendable () -> Date = { Date() }) {
        self.state = state
        self.sealedAtBoot = sealedAtBoot
        self.clock = clock
        self.stateChangedAt = clock()
    }

    /// Apply one event. Returns the effects to execute; throws on an event
    /// that is meaningless in the current state (a programming error or a
    /// stale caller — never silently swallowed).
    @discardableResult
    public mutating func handle(_ event: VMEvent) throws -> [VMEffect] {
        let (next, effects) = try transition(for: event)
        state = next
        stateChangedAt = clock()
        return effects
    }

    private func transition(for event: VMEvent) throws -> (VMState, [VMEffect]) {
        switch (state, event) {
        case (.created, .startInstall):
            return (.installing, [.attachInstallerISO, .startVirtualMachine])
        case (.failed(let f), .startInstall) where f.phase == .install:
            return (.installing, [.attachInstallerISO, .startVirtualMachine])

        case (.installing, .installSucceeded):
            // The install→first-boot seam: the ISO comes OFF here so every
            // subsequent boot is from the guest's own disk.
            return (.installed, [.detachInstallerISO])
        case (.installing, .installFailed(let reason)):
            return (.failed(VMFailure(phase: .install, reason: reason)),
                    [.stopVirtualMachine, .detachInstallerISO])

        case (.installed, .powerOn), (.stopped, .powerOn):
            return (sealedAtBoot ? .awaitingPhoneUnlock : .running, [.startVirtualMachine])
        case (.failed(let f), .powerOn) where f.phase == .run:
            return (sealedAtBoot ? .awaitingPhoneUnlock : .running, [.startVirtualMachine])

        case (.awaitingPhoneUnlock, .guestUnlocked):
            return (.running, [])
        case (.awaitingPhoneUnlock, .powerOff), (.running, .powerOff):
            return (.stopped, [.stopVirtualMachine])
        case (.awaitingPhoneUnlock, .runtimeFailed(let reason)),
             (.running, .runtimeFailed(let reason)):
            return (.failed(VMFailure(phase: .run, reason: reason)), [])

        default:
            throw VMLifecycleError.invalidTransition(from: state, on: event)
        }
    }
}

// MARK: - Install-stop interpretation (the reboot/poweroff/never-booted seam)

extension VMLifecycle {
    /// The minimum wall-clock time from entering `.installing` for a CLEAN
    /// (no-error) guest self-stop to plausibly mean "the unattended install
    /// finished". A real Debian-preseed install takes many minutes; a clean
    /// stop faster than this means the guest did NOT install:
    ///
    ///   - it failed to boot at all — e.g. an amd64 base image on an
    ///     Apple-silicon host, which VZ cannot run: the VM stops ~0.3s after
    ///     start with `error=nil` (observed during Phase-0 hardware bring-up);
    ///   - or the installer rebooted very early (a mid-install failure).
    ///
    /// The seam must NOT read such a fast clean stop as success, or a VM that
    /// never ran an install is reported "Installed" and then boots an empty
    /// disk. Only a clean stop after a plausible duration is a real install
    /// completion — and that is true whether the finished installer POWERED OFF
    /// or REBOOTED (VZ surfaces both as a delegate `guestDidStop`, so the
    /// distinction is duration, not the stop kind).
    public static let minPlausibleInstallDuration: TimeInterval = 90

    public enum InstallStopVerdict: Equatable, Sendable {
        case installed
        case failedTooFast(elapsed: TimeInterval)
    }

    /// Interpret a clean guest self-stop observed while `.installing`.
    /// `installStartedAt` is when `.installing` was entered (this lifecycle's
    /// `stateChangedAt`); `now` from the caller's clock.
    public static func verdictForCleanInstallStop(installStartedAt: Date,
                                                  now: Date) -> InstallStopVerdict {
        let elapsed = now.timeIntervalSince(installStartedAt)
        return elapsed >= minPlausibleInstallDuration
            ? .installed
            : .failedTooFast(elapsed: elapsed)
    }

    /// A sealed guest awaiting phone-unlock should come online within a few
    /// minutes; past this it has very likely failed to reach the network (e.g. a
    /// first-boot NIC/DHCP failure) and would otherwise spin on "Waiting for you
    /// to unlock" forever with no hint. The UI keeps polling, but past this
    /// threshold it surfaces an advisory instead of an indefinite spinner.
    public static let comingUpStallThreshold: TimeInterval = 8 * 60

    /// True iff a hosted server has been sealed + awaiting unlock past the stall
    /// threshold. Pure so the view can evaluate it against a live `now`.
    public static func comingUpIsStalled(state: VMState, elapsed: TimeInterval) -> Bool {
        state == .awaitingPhoneUnlock && elapsed >= comingUpStallThreshold
    }
}

File: apps/builder-windows/src/VM/VMLifecycle.cs
using System;
using System.Collections.Generic;

namespace Flagship.Builder.VM;

/// <summary>
/// One hosted VM's lifecycle — pure and event-driven. The host (QEMU adapter /
/// UI) feeds it events and executes the effects it returns; nothing in here
/// touches QEMU, the filesystem, or a real clock.
///
///     created ──startInstall──▶ installing (installer ISO attached)
///     installing ──installSucceeded──▶ installed (ISO DETACHED — from here
///                                       the guest boots from its own disk)
///     installed/stopped ──powerOn──▶ awaitingPhoneUnlock   (encrypted guest:
///                                       sealed in the initramfs until the
///                                       phone-home unlock answers)
///                              └────▶ running              (unencrypted guest)
///     awaitingPhoneUnlock ──guestUnlocked──▶ running
///     running/awaitingPhoneUnlock ──powerOff──▶ stopped
///     + failure states (install / run), each retryable.
///
/// Mirrors apps/builder-mac FlagshipBuilderCore/VM/VMLifecycle.swift; the
/// shared contract is pinned by apps/desktop-shared/golden/vm-core-vectors.json.
/// </summary>
public enum VMFailurePhase { Install, Run }

public sealed record VMFailure(VMFailurePhase Phase, string Reason);

public enum VMStateKind
{
    Created,
    Installing,
    Installed,
    AwaitingPhoneUnlock,
    Running,
    Stopped,
    Failed,
}

public sealed record VMState(VMStateKind Kind, VMFailure? Failure = null)
{
    public static readonly VMState Created = new(VMStateKind.Created);
    public static readonly VMState Installing = new(VMStateKind.Installing);
    public static readonly VMState Installed = new(VMStateKind.Installed);
    public static readonly VMState AwaitingPhoneUnlock = new(VMStateKind.AwaitingPhoneUnlock);
    public static readonly VMState Running = new(VMStateKind.Running);
    public static readonly VMState Stopped = new(VMStateKind.Stopped);
    public static VMState Failed(VMFailure failure) => new(VMStateKind.Failed, failure);
    public static VMState Failed(VMFailurePhase phase, string reason) => new(VMStateKind.Failed, new VMFailure(phase, reason));

    /// <summary>User-facing status label (sidebar / detail).</summary>
    public string Label => Kind switch
    {
        VMStateKind.Created => "Created",
        VMStateKind.Installing => "Installing…",
        VMStateKind.Installed => "Installed",
        VMStateKind.AwaitingPhoneUnlock => "Waiting for you to unlock",
        VMStateKind.Running => "Running",
        VMStateKind.Stopped => "Stopped",
        VMStateKind.Failed => Failure?.Phase == VMFailurePhase.Install ? "Install failed" : "Stopped unexpectedly",
        _ => "Unknown",
    };
}

public enum VMEventKind
{
    StartInstall,
    InstallSucceeded,
    InstallFailed,
    PowerOn,
    GuestUnlocked,
    PowerOff,
    RuntimeFailed,
}

public sealed record VMEvent(VMEventKind Kind, string? Reason = null)
{
    /// <summary>Begin (or retry) the unattended install from the remastered ISO.</summary>
    public static readonly VMEvent StartInstall = new(VMEventKind.StartInstall);
    /// <summary>The installer ran to completion (the guest powered itself off).</summary>
    public static readonly VMEvent InstallSucceeded = new(VMEventKind.InstallSucceeded);
    public static VMEvent InstallFailed(string reason) => new(VMEventKind.InstallFailed, reason);
    /// <summary>Boot the installed guest from its main disk.</summary>
    public static readonly VMEvent PowerOn = new(VMEventKind.PowerOn);
    /// <summary>The sealed guest's phone-home unlock completed (owner approved /
    /// a lease answered) and the guest is up.</summary>
    public static readonly VMEvent GuestUnlocked = new(VMEventKind.GuestUnlocked);
    /// <summary>The guest stopped (user action or a clean guest shutdown).</summary>
    public static readonly VMEvent PowerOff = new(VMEventKind.PowerOff);
    public static VMEvent RuntimeFailed(string reason) => new(VMEventKind.RuntimeFailed, reason);
}

/// <summary>
/// Side effects the caller must perform, in order. The state machine decides;
/// the QEMU layer obeys.
/// </summary>
public enum VMEffect
{
    AttachInstallerISO,
    DetachInstallerISO,
    StartVirtualMachine,
    StopVirtualMachine,
}

public sealed class VMLifecycleException : Exception
{
    public VMState From { get; }
    public VMEvent On { get; }

    public VMLifecycleException(VMState from, VMEvent on)
        : base($"Invalid VM transition: {from.Kind} on {on.Kind}.")
    {
        From = from;
        On = on;
    }
}

public sealed class VMLifecycle
{
    public VMState State { get; private set; }
    /// <summary>When the current state was entered (via the injected clock).</summary>
    public DateTimeOffset StateChangedAt { get; private set; }

    /// <summary>
    /// From VMConfig.AwaitsPhoneUnlockAtBoot: whether powering on an installed
    /// guest lands in the sealed awaiting-phone-unlock state first.
    /// </summary>
    public bool SealedAtBoot { get; }

    private readonly Func<DateTimeOffset> _clock;

    public VMLifecycle(bool sealedAtBoot, VMState? state = null, Func<DateTimeOffset>? clock = null)
    {
        State = state ?? VMState.Created;
        SealedAtBoot = sealedAtBoot;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        StateChangedAt = _clock();
    }

    /// <summary>
    /// A hypervisor-level clean guest-stop during the install phase is
    /// AMBIGUOUS: install-success (the preseed powers the guest off), a
    /// completed-install reboot, and a never-really-booted guest all look
    /// identical from outside. The Mac Phase-0 boot proved the only reliable
    /// discriminator is elapsed time — a real unattended Debian install cannot
    /// complete faster than this.
    /// </summary>
    public static readonly TimeSpan MinPlausibleInstallDuration = TimeSpan.FromSeconds(90);
    public static readonly TimeSpan MinPlausibleSpecializationDuration = TimeSpan.FromSeconds(5);

    /// <summary>
    /// The duration-gated verdict for a clean guest-stop observed while
    /// installing: a clean stop after a plausible duration is success (poweroff
    /// OR reboot — both mean the installer finished); a too-fast clean stop is
    /// a failure with an actionable message. Pure — pinned by the shared
    /// golden vectors (edge: exactly the minimum counts as success).
    /// </summary>
    public static VMEvent VerdictForCleanInstallStop(TimeSpan elapsedSinceInstallStart)
    {
        if (elapsedSinceInstallStart >= MinPlausibleInstallDuration) return VMEvent.InstallSucceeded;
        var secs = (int)Math.Round(elapsedSinceInstallStart.TotalSeconds);
        return VMEvent.InstallFailed(
            $"The installer stopped after only {secs}s — too fast to have completed. " +
            "The guest likely failed to boot the installer ISO. Retry the install; " +
            "if it persists, re-download the base image and remaster again.");
    }

    public static VMEvent VerdictForCleanProvisioningStop(
        TimeSpan elapsedSinceInstallStart, VMProvisioningMode mode)
    {
        if (mode == VMProvisioningMode.InstallerISO)
            return VerdictForCleanInstallStop(elapsedSinceInstallStart);
        if (elapsedSinceInstallStart >= MinPlausibleSpecializationDuration)
            return VMEvent.InstallSucceeded;
        var secs = (int)Math.Round(elapsedSinceInstallStart.TotalSeconds);
        return VMEvent.InstallFailed(
            $"The appliance stopped after only {secs}s — too fast to have specialized. " +
            "The image may be the wrong architecture or otherwise not bootable.");
    }

    /// <summary>
    /// Convenience for the live adapter: derive the elapsed install time from
    /// this machine's own state timestamp. Only meaningful while installing.
    /// </summary>
    /// <summary>
    /// A sealed guest awaiting phone-unlock should come online within a few
    /// minutes; past this it has very likely failed to reach the network (e.g. a
    /// first-boot NIC/DHCP failure) and would otherwise spin on "Waiting for you
    /// to unlock" forever with no hint. The UI keeps polling, but past this
    /// threshold it surfaces an advisory. Mirrors macOS
    /// VMLifecycle.comingUpStallThreshold / Linux COMING_UP_STALL_THRESHOLD.
    /// </summary>
    public static readonly TimeSpan ComingUpStallThreshold = TimeSpan.FromMinutes(8);

    /// <summary>
    /// True iff a hosted server has been sealed + awaiting unlock past the stall
    /// threshold. Pure so the view can evaluate it against a live clock.
    /// </summary>
    public static bool ComingUpIsStalled(VMStateKind state, TimeSpan elapsed) =>
        state == VMStateKind.AwaitingPhoneUnlock && elapsed >= ComingUpStallThreshold;

    public VMEvent VerdictForCleanInstallStop(DateTimeOffset now)
    {
        if (State.Kind != VMStateKind.Installing)
            throw new InvalidOperationException($"Install verdict requested in state {State.Kind}.");
        return VerdictForCleanInstallStop(now - StateChangedAt);
    }

    /// <summary>
    /// Apply one event. Returns the effects to execute; throws on an event
    /// that is meaningless in the current state (a programming error or a
    /// stale caller — never silently swallowed).
    /// </summary>
    public IReadOnlyList<VMEffect> Handle(VMEvent ev)
    {
        var (next, effects) = Transition(ev);
        State = next;
        StateChangedAt = _clock();
        return effects;
    }

    private (VMState, VMEffect[]) Transition(VMEvent ev)
    {
        switch (State.Kind, ev.Kind)
        {
            case (VMStateKind.Created, VMEventKind.StartInstall):
                return (VMState.Installing, new[] { VMEffect.AttachInstallerISO, VMEffect.StartVirtualMachine });
            case (VMStateKind.Failed, VMEventKind.StartInstall) when State.Failure?.Phase == VMFailurePhase.Install:
                return (VMState.Installing, new[] { VMEffect.AttachInstallerISO, VMEffect.StartVirtualMachine });

            case (VMStateKind.Installing, VMEventKind.InstallSucceeded):
                // The install→first-boot seam: the ISO comes OFF here so every
                // subsequent boot is from the guest's own disk.
                return (VMState.Installed, new[] { VMEffect.DetachInstallerISO });
            case (VMStateKind.Installing, VMEventKind.InstallFailed):
                return (VMState.Failed(VMFailurePhase.Install, ev.Reason ?? ""),
                        new[] { VMEffect.StopVirtualMachine, VMEffect.DetachInstallerISO });

            case (VMStateKind.Installed, VMEventKind.PowerOn):
            case (VMStateKind.Stopped, VMEventKind.PowerOn):
                return (SealedAtBoot ? VMState.AwaitingPhoneUnlock : VMState.Running,
                        new[] { VMEffect.StartVirtualMachine });
            case (VMStateKind.Failed, VMEventKind.PowerOn) when State.Failure?.Phase == VMFailurePhase.Run:
                return (SealedAtBoot ? VMState.AwaitingPhoneUnlock : VMState.Running,
                        new[] { VMEffect.StartVirtualMachine });

            case (VMStateKind.AwaitingPhoneUnlock, VMEventKind.GuestUnlocked):
                return (VMState.Running, Array.Empty<VMEffect>());
            case (VMStateKind.AwaitingPhoneUnlock, VMEventKind.PowerOff):
            case (VMStateKind.Running, VMEventKind.PowerOff):
                return (VMState.Stopped, new[] { VMEffect.StopVirtualMachine });
            case (VMStateKind.AwaitingPhoneUnlock, VMEventKind.RuntimeFailed):
            case (VMStateKind.Running, VMEventKind.RuntimeFailed):
                return (VMState.Failed(VMFailurePhase.Run, ev.Reason ?? ""), Array.Empty<VMEffect>());

            default:
                throw new VMLifecycleException(State, ev);
        }
    }
}
